alf/algorithms/algorithm.py

```diff
@@ -1533,6 +1714,214 @@ def _update(self, experience, batch_info, weight):
+                experience, self.experience_spec, batch_info,
+                mini_batch_length, self._replay_buffer)
+
+            if self._processed_experience_spec is None:
+                self._processed_experience_spec = dist_utils.extract_spec(
+                    experience, from_dim=2)
+
+        # TODO: use a different mini_batch_length for aux
+        offline_experience, offline_batch_info, length, mini_batch_length, batch_size = self._prepare_experience_data(
+            offline_experience, self._offline_experience_spec,
+            offline_batch_info, mini_batch_length, self._offline_replay_buffer)
+
+        def _make_time_major(nest):
+            """Put the time dim to axis=0."""
+            return alf.nest.map_structure(lambda x: x.transpose(0, 1), nest)
+
+        indices = None
+        for u in range(num_updates):
+            if mini_batch_size < batch_size:
+                # here we use the cpu version of torch.randperm(n) to generate
+                # the permuted indices, as the cuda version of torch.randperm(n)
+                # seems to have a bug when n is a large number, generating
+                # negative or very large values that cause out of bound kernel
+                # error: https://github.com/pytorch/pytorch/issues/59756
+                indices = alf.nest.utils.convert_device(
+                    torch.randperm(batch_size, device='cpu'))
+
+            for b in range(0, batch_size, mini_batch_size):
+                if update_counter_every_mini_batch:
+                    alf.summary.increment_global_counter()
+
+                is_last_mini_batch = (u == num_updates - 1
+                                      and b + mini_batch_size >= batch_size)
+                do_summary = (is_last_mini_batch
+                              or update_counter_every_mini_batch)
+                alf.summary.enable_summary(do_summary)
+
+                if indices is None:
+                    batch_indices = slice(b,
+                                          min(batch_size, b + mini_batch_size))
+                else:
+                    batch_indices = indices[b:min(batch_size, b +
+                                                  mini_batch_size)]
+
+                if experience is not None:
+                    batch = alf.nest.map_structure(lambda x: x[batch_indices],
+                                                   experience)
+                    if batch_info:
+                        binfo = alf.nest.map_structure(
+                            lambda x: x[batch_indices] if isinstance(
+                                x, torch.Tensor) else x, batch_info)
+                    else:
+                        binfo = None
+                    batch = _make_time_major(batch)
+                    batch = alf.nest.utils.convert_device(batch)
+                else:
+                    batch = None
+                    binfo = None
```

Comment:
This major code block `updates x mini-batches` looks very similar to the one in `_train_experience`. Maybe we can extract the common logic to decrease the maintenance effort in the future.

Reply:
Now more code sharing with a shared function.